### ISSUE TITLE
Proposal: enhance ui_link and ui_links_row to pass hints/classes for theme

### ISSUE THREAD

**Issue by gnadelwartz** [webmin/webmin#678]
in the firewall module janie [accepted](https://github.com/webmin/webmin/pull/676) a "hack" from me to visually improve the links to switch between IPv4 and IPv6. IMHO this should be done by theme.

A possible way to make it easyer for the theme author is to allow passing hints (classes) in ui_link_row.
today all classes from a link `<a class="danger" href="...">xxx</a>` are changed to `<a class="ui-link" href="...">xxx</a>`.

we can use a defined ste of classes e.g. success, danger, warning, info etc. like in bootstrap.
In the firewall  example the code can look like this:

ui_link_row(['<b class="info">IPv4</b>','<a class="primary" href="xxx">IPv6</a>']);

**Comment by jcameron** [comment]
Interesting idea ... but what would the hints be used for? Coloring the links?

**Comment by gnadelwartz** [comment]
clororing and everithing else the theme author likes todo :-) e.g. today Ilja converts ul-link to a button. with this hint he can also set a bgcolor for info.

the general idea comes from my recent use of bootstrap and is not limited to ui_link. Since Ilja already includex bootstrap, we have the possibility to easyly use it from webmin. Other themes without bootstrap can define their own css if they want to support it, if theme does not kown about its dislayed like today.

e.g. we can hint the firewall warnings with "alert alert-danger" and authentic display it automatically this way:
![grafik](https://user-images.githubusercontent.com/4593242/32406608-cb56ae04-c17b-11e7-87f0-ff623602c473.png)
```
print "<div class=\"alert alert-danger\">",
                   &text("index_$word", "$gconfig{'webprefix'}/$word/", $fwconfig),
                   "</div><p>\n";
```
Grey Framed theme:
![grafik](https://user-images.githubusercontent.com/4593242/32406654-87279a6c-c17c-11e7-9f4c-513e551bb6cc.png)

as you can see its backwards compatible, so it may be a nice extentsion.

**Comment by qooob** [comment]
@gnadelwarty Kay, ideally we should separate logic and presentation completely. 

Have a look at `theme_ui_alert_box` sub, just do what you do using it like

```Perl
theme_ui_alert_box('It appears that Shorewall is ....', 'danger')
```

You should get your alert box.

**Comment by qooob** [comment]
My bad, you shouldn't call `theme_ui_alert_box` but `ui_alert_box`.

**Comment by gnadelwartz** [comment]
Nevertheless supporting all bootsrap classes is to much. I you like it we schould allow only a small subset.

- text: https://www.w3schools.com/bootstrap/bootstrap_ref_css_helpers.asp
![grafik](https://user-images.githubusercontent.com/4593242/32406812-72d915e2-c17f-11e7-9c90-dc2dd987882a.png)
- alert: https://www.w3schools.com/bootstrap/bootstrap_alerts.asp
![grafik](https://user-images.githubusercontent.com/4593242/32406824-95ee8d14-c17f-11e7-9bd9-fa8d33ad58f0.png)
- buttons: https://www.w3schools.com/bootstrap/bootstrap_buttons.asp 
![grafik](https://user-images.githubusercontent.com/4593242/32406829-a75e9f80-c17f-11e7-9aa6-dfd17443e1d7.png)

**Comment by qooob** [comment]
Kay, I'm not sure what you're asking? Could you rephrase your question?

**Comment by gnadelwartz** [comment]
@iliaross yes, thas the way we should implement it in webmin. is this call from your theme or from webmin?
This is also why my proposal was to extend the webmin UI_xxx calls to support this type of hints, so not every theme author has to overwrite the ui_functions for his own.

**Comment by qooob** [comment]
It's provided by Webmin `ui-lib.pl`, that checks for `theme_ui_alert_box`, in case it's not defined the default used:

[ui-lib.pl#L2387-L2425](https://github.com/webmin/webmin/blob/master/ui-lib.pl#L2387-L2425)
```Perl

=head2 ui_alert_box(msg, type)

Returns HTML for an alert box, with background color determined by $type.

$msg contains any text or HTML to be contained within the alert box, and
can include forms.

Type of alert:

=item success - green

=item info - blue

=item warn - yellow

=item danger - red

=cut

sub ui_alert_box
{
my ($msg, $type) = @_;
my ($rv, $color);

if (defined (&theme_ui_alert_box)) {
    return &theme_ui_alert_box(@_);
    }

if ($type eq "success") { $color = "DFF0D8"; }
elsif ($type eq "info") { $color = "D9EDF7"; }
elsif ($type eq "warn") { $color = "FCF8E3"; }
elsif ($type eq "danger") { $color = "F2DEDE"; }

$rv .= "<table class='ui_alert_box' width='100%'><tr bgcolor='#$color'><td align='center'><p>\n";
$rv .= "$msg\n";
$rv .= "<p></td></tr></table><p>\n";

return $rv;
}
```

**Comment by qooob** [comment]
> This is also why my proposal was to extend the webmin UI_xxx calls to support this type of hints, so not every theme author has to overwrite the ui_functions for his own.

This means making a dependency and worse code, overall.

**Comment by gnadelwartz** [comment]
@iliaross OK, after your Input I would say it this way: For Webmin 2.0 we should consider moving some of your theme specific ui_functions to webmin core api :-) in a backwards compatible way.

**Comment by qooob** [comment]
I would completely strip of Webmin UI library from any HTML and moved it to a very basic theme (shipped with Webmin 2.0).

**Comment by gnadelwartz** [comment]
so all ui_functions are provided by the theme and any HTML has to be removed from CGI scripts?

**Comment by qooob** [comment]
UI lib has basic functions that can be overridden by the theme. It's a library and it has HTML in it. As long as it's the only file containing it, I think it's alright.

Building both functional and properly architectured system is very difficult task. What makes things even more difficult is that project's development started over 20 years ago.

The point of Webmin 2.0 branch is to build contemporary app, that would be an example in terms of codebase. Authentic Theme 20 is going to aspire the same goals. It's just so much work to do.

**Comment by swelljoe** [comment]
I disagree with Ilia on a couple of points above. I think the ideal would allow for modules to mark the type of botton or alert or whatever it is; this could be the Bootstrap default names, since they're popular and we're already using Bootstrap in the Authentic theme (and I may use it in the new Webmin "no-theme theme", though I might use something lighter).

But, where I disagree is with color being defined in code rather than stylesheets (that should just attach the class "danger" or whatever to the element being generated instead of hardcoding colors in Perl code).

And I also disagree with the notion that Webmin ui-lib.pl shouldn't be able to handle this. If we are using color to convey information (hopefully not in a way that makes colorblind people miss that information, just as augmentation), then the core Webmin should be aware of it and know to do something with it...even if it just means attaching a class to the elements (which is probably what all themes should do...and that means that maybe themes that currently *must* override the ui-lib functions won't have to do so anymore for the majority of ui elements).

I want to see themes *easier* to make, not more flexible to make. Flexibility should come from spitting out good, validated HTML, and clear CSS classes and IDs on every element. So, a web designer should be able to look at what Webmin produces, and make a theme entirely with a couple of HTML templates and a CSS files. Ideally we'd be able to move to a client-side app by the time of Webmin 3.0, where the client receives JSON objects or GraphQL query data and draws a UI using something like React or Vue.js. But, we've got a lot of work to do to get there.

But, my goal in Webmin 2.0 is to make it so a theme can be *much* smaller than currently. I hate how large Authentic has to be to be so awesome. At the very least, the ui-lib ought to be able to spit out HTML that works for Authentic and any other theme like it. You shouldn't have to override ui-lib functions in the general case, IMHO. The client-side code (JavaScript and CSS and HTML) will remain large, but those things are comprehensible to web designers. The back-end Perl isn't (and I think the number of people willing to work with Perl is in decline, unfortunately).

**Comment by chris001** [comment]
Right on.  The back end should be its own app and can be coded in perl or anything readily available on linux servers, and shouldn't deal with HTML, only data.  The front end should be a separate app, which makes calls to the back end app, to create, read, update and delete data for the current screen, and displays it in responsive css html.  The two apps talk to one another through a standard text based API like JSON or graphQL.

**Comment by qooob** [comment]
I also stand for having colors being defined in stylesheet rather than in Perl. I don't remember proposing otherwise.

Perl and JavaScript have a lot in similar, just as any other contemporary programming languages. It's not a problem for any programmer to figure things out by looking at the code. It doesn't matter what programming language is it, the idea is always the same, though, the syntax is different.

What would be the best thing to do, in my opinion, is to completely get rid of UI library by just sending to a client (theme) prestructured JSON object with all needed data directly by modules. On its side, theme has to build its own UI engine and decided how it should be drawn/done. Then we separate logic and presentation completely. Then it will be easy to build an Android/iOS app for example, just by following the same principals. Returning HTML is always bad.

I think exactly that should be our goal for Webmin 2.0. Having structural changes, only visible to programmers, is only fun for us. Most users wouldn't give a thing of how beautiful the code is. As a developer (client or server - it's irrelevant), I want to have an API returning data I could easily manipule and work with. I wouldn't want to deal with somebody else's logic of how it should be done. The point is that. Besides, it's very important to leave the opportunity for themes to run any server code (by switching to needed user), for building fun and really extensible app (theme)). There is no way to build anything close as functional as Authentic without having this ability.

>The client-side code (JavaScript and CSS and HTML) will remain large, but those things are comprehensible to web designers. The back-end Perl isn't.

Joe, nowadays, the line between client and server side is interflowed. I don't consider myself to be a web designer but rather a programmer. Without doing server side all a developer could do in the end is boring and very little functional app.

> I want to see themes easier to make, not more flexible to make.

On doing that we would get more themes in number but less in terms of quality and usability. 

----

I believe, we should never even think close of limiting anybody, neither users nor especially developers. 

I think we should learn from Google. Let's say from Android in this regard. It's super powerful and has tons of features, compared to iOS. Android doesn't force you to go through all of its options but leaves you such opportunity. It's not enforcing.

We don't need to make oversimplified (entry-level) application in thinking that some users just won't be able to figure things out. All we need is just to make things work out of the box and provide endless potential for those users and developers who are willing to go deeper.

That was the only reason why I being a newbie, chosen Webmin/Virtualmin over cPanel/Plesk and paid for the full license (never actually needing one).

Using an app should leave you the feeling, that there is something bigger left for you. Most importantly that this app is really made and can be tweaked for/by _you_ and that is impossible in terms of raising simplicity.

**Comment by gnadelwartz** [comment]
I think you are both rigth. Customizing (Theme) of webmin should be simple and to we should move webmin from last century to a modern world (authetic). The basics and flexibility are already there as Ilja showed us with his famous authemtic-theme. To achive both goal we may need a step by step aproach.

- replace all HTML/CSS stuff from scripts with call's to ui_lib, possibly add hints for presentation layer
- build a new simple basic theme using HTML/CSS for presentation, theme CSS can be easily overloaded for customizing.
- app's (and I have to call Ilja's work an app) can overload ui_lib to use a modern/new/whatever presentation layer, which does not have to care about the crappy HTML from scripts as today

![grafik](https://user-images.githubusercontent.com/4593242/32413840-126083d0-c21a-11e7-8750-30b480f73f3c.png)

**Comment by gnadelwartz** [comment]
Since jamie was not against it, joe supported hints and Ilia seem to like my last comment (oh chris also :), I spend some time to investigate and think about it.

Besides what hints may be useful, I found two challenges with to get rid of HTML in scripts:
1. the UI API is not designed to replace HTML, its more about harmonize UI Design  
2. its explicitly stated to use HTML in the API description. 

Thats not bad, because HTML markup can also be used to structure data, but It may necessary to limit the accepted tags. I preperared a first draft what hints and tags may be useful.

Should I post it here or do you have a better place for ducuissing things like this? e.g. internal forum, slack, teamroom ...

**Comment by gnadelwartz** [comment]
@swelljoe you mentioned to use bootstrap for the new theme, but it may be to havey.

I looked for bootstrap alternatives, but it seems like the good one are based on bootstrap and the ligth ones has only selected use cases and/or have completly different classes.

A compromise may be to include only the [CSS file](https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.css) from bootstrap and ignore the fancy JS stuff.

**Comment by gnadelwartz** [comment]
**Hints**

If you're already using a  modern framework like *bootstrap* you will notice that webmin hints are very similar to bootstrap classes. The main difference is that hints are more general and context sensitive.

Bootstrap has many similar classes for elements plus some general classes:

```
- text-primary, btn-info,  alert-success, ...
- active, success, info, warning, danger
```

In webmin only general hints should be used because the context (element type) is already known by the name of the function call.

The following hints may be accepted from every ui_lib function. How they are interpreted depends on ui_lib function, theme or app.

```
- active
- inactive (disabled, muted, invisible)
- primary
- success
- info
- warning
```

some functions may allow additonal arguments, e.g. tables, button_row

```
- striped
- condensed
- nobreak
- left, rigth, center
```


**HTML markup in text/data**

In old webmin you can use any HTML markup direkt from print statement or text/data provided to ui_lib functions. With webmin 2.0 only a restricted set of markup should be used:

```
- <h1> … <h6>
- <p>, </br>
- <small>
- <strong> (<b>)
- <mark>
- <abbr>
- <blockquote>
- <code> (<pre>, <tt>)
- <kbd>
- ...
```

**Possible Usage**

Hints may provided as optional argument to ui_lib functions. If a function already has a formating argument the hints may be be used there. possible examples:

```
- ui_alert_box(text, hints)
- ui_table_start(heading, [hints], [cols], [&td-hints], [right-heading])
- ui_table_row(label, value, [cols], [&td-hints])
- ui_table_span(text, hint)
- ui_button(label, [name], [disabled?], [hints])
- ui_links_row(&links, [&hints])
```

These are only examples how it *may* look, it depends on how ui_lib is refactored for webmin 2.0

**Comment by gnadelwartz** [comment]
**Special Hints**

I was also thinking about special hints for inputs and buttons to provide ui_lib information about the resulting action, e.g. its a delete/add/up/down/insert/save etc. But I'm not shure if this makes sense.

```
- add
- delete
- insert
- up, down
- save
- open
- OK
- cancel
- reset
- ...
```

**Comment by qooob** [comment]
Kay, that would require to do this for each occurrence in each module's file, meaning thousands of it. It's a lot of work for cosmetic changes.

While Joe is being more realistic, and returning valid HTML without tons of nested tables would already be a huge progress. Doing it likely possible for the next Webmin branch rather than switching completely to JSON objects for page rendering.

The main problem for HTML, is that it has to be parsed. It's not a problem for completely valid HTML. Returning valid HTML would already be a big step forward. My idealistic approach in switching to JSON objects right now may not be attainable, and I realize it.

**Comment by qooob** [comment]
We have Slack. It's not publicly opened yet, thus requires invite from our side.

You can write me if you want to be added to `authentic-theme.slack.com` or Eric to be added to `virtualmin.slack.com`.

**Comment by gnadelwartz** [comment]
Modules has to be edited anyway to convert all print statements containing plain HTML Markup over to ui_lib calls, this is indeed a huge work. But without this seems impossible to get validated HTML output.

If all these plain `table, form, input` tags are converted to ui_lib calls, the remaining HTML in ui_lib calls already fit to the restricted HTML set.

We may end then with a `Module API` of ui_lib calls with restricted HMTL and optional hints. ui_lib itself spit out validated HTML with classes and IDs attached. lets call this `Presentation API`.

If webmin in future want's use `JSON` for the presentation layer it's "only" a matter of overload/replace ui_lib to spit out `JSON`, the modules can stay as is because the `Module API` doesn't change.

Adavantages this step by step approach are:

- module developers can stick with perl, ui_lib calls and HTML
- you can convert (external) modules step by step
- it's possible to have multiple Frontends

I updated my picture in my comment above to refelect this.

Disclaimer: As noted in our private chat, I'm an old style developer using like perl and plain HTML. Modern Stuff like JSON, AJAX etc. is evil magic ;-)

**Comment by qooob** [comment]
Majority of modules could be converted with ease, more or the less, as UI library used pretty much everywhere.

More complex is to properly replace nested tables with grid system, using `col-` classes to properly display the page. Ideally table should be used once on the page and only for storing table-like data, not for containing checkboxes, radios, inputs and other. This is harder to do than it seems.

Yes, it's still going to require to edit modules.

I was proposing to go directly to JSON as I see already so much work to do. With JSON it's even more but, _maybe_, it's worth doing it straight away now, right to how it should be ideally than putting this off for the future.

Again maybe, we could just start the project from the ground and use as the main framework Mojolicious for instance and import all modules and rest of the code step by step. Even more work but that would take us closer to where we want to get.

It's all about users and their support. It's very difficult task for practically only Joe to accomplish. Jamie, I personally, don't think will be able to invest that much time. I could also provide help in this regard but in parallel need to continue development of the theme.

At the end, it's up-to Jamie and Joe to decide. They are smarter (than I am) in server-side scripting and will definitely be able to make the right, better choice.

**Comment by gnadelwartz** [comment]
full ack, and I'm pretty shure Jamie and Joe will come up with the right decision how to go forward after the dust of our small discussion has settled and their popcorn is empty 😋😁

**Comment by gnadelwartz** [comment]
to get an idea how much work has to be done to remove HTML from a module I started with my reworked mailman module. https://github.com/webmin/webmin/pull/684

holy s**t, that's damn hard work! If there are more modules with this amount of HTML to convert it will take years!

**Comment by qooob** [comment]
Unfortunately, yes, there are a lot of work. This is why I create global patches to make things work with certain patterns.

It's not ideal though.

**Comment by chris001** [comment]
How about using a modern IDE, like JetBrains WebStorm EAP free download, with the free Camelcade Perl plugin.  It could auto refactor these HTML to function call.
https://confluence.jetbrains.com/display/WI/WebStorm+EAP
https://plugins.jetbrains.com/plugin/7796-perl
Also Webmin contributors can get a free license for WebStorm as an open source project.

**Comment by swelljoe** [comment]
It's not possible to automatically convert any of this stuff, no matter how clever the tool. The semantics of tables and CSS grids are just different, and there's no automatic way to convert, e.g. print "some random HTML" into ui-lib calls. The reality is, some person (or hopefully multiple persons), will sit down with the modules that don't yet fully use ui-lib and convert them. It's just a big tedious (though not terribly difficult) job. Using an IDE won't change that. Tackling one or more modules would be a reasonable first project for someone wanting to get into Webmin development. It's not technically very hard, it just requires a bit of trial and error and a little HTML and Perl knowledge.

Making ui-lib more modern is do-able by a single person in a couple of weeks...I've mostly done it before (for my Bootstrap theme that I gave up on when Authentic came along) and Ilia has mostly replaced quite a bit of ui-lib functions, as well, for Authentic. But, unless/until every module we ship in Webmin uses ui-lib consistently and in ways that don't break HTML (nested forms is a big problem in a lot of places that leads to invalid HTML that can't be parsed reliably with JavaScript). With my previous single-page app theme attempt, I ran into a bunch of problems with invlid HTML; at the time, there were a lot more modules not using ui-lib (Jamie has been improving the ui-lib coverage for years now...he's done more of these conversions than anyone) so it was just too frustrating to complete. But, we're probably at the point where I could make the conversion in Webmin 2.0 branch and just fix problems as they're found.

**Comment by swelljoe** [comment]
To be clear: even modules that currently use ui-lib today may be using it in a way that produces invalid HTML which also prevents a single-page app model from working (and most of the current bugs in the new 19.x version of Authentic are variants of this problem). So, we have to both convert all modules to ui-lib *and* fix any modules that use ui-lib in ways that generate invalid HTML. This is a lot of little tweaks, but no major architectural changes are needed.

**Comment by gnadelwartz** [comment]
may be we can add some simple checks to ui-lib to report e.g. nested forms or not closed tables.
this nay help to detect sources if invalid html

**Comment by qooob** [comment]
@gnadelwartz That's not gonna be practically useful, I'm afraid.

**Comment by chris001** [comment]
Actually @iliaross , what @gnadelwartz says sounds sensible.  By adding some simple HTML validity/sanity/purity/tidy checks to `ui-lib` , it could log its findings, for example, "invalid HTML returned from module ____ function ____ input: ____ output: ____", it becomes a time saving tool, to pinpoint which modules and which APIs inside them are returning broken/invalid/deprecated HTML, nested tables, etc. Isn't it?

**Comment by qooob** [comment]
@chris001 Probably, but nearly each module has HTML output issues. From misplaced forms, and up to having multiple unsupported attributes on the tags.

It would worth doing it after the UI rewrite for Webmin 2.0, I think.

Having complete W3C validation, as HTML5 is doable but very challenging for us.

**Comment by chris001** [comment]
Yes, passing W3C validation HTML5 will make it one simple step more to a responsive CSS3 UI which lets the user's browser automatically adapt the font size and page element layout, to enable you to use cloudmin virtualmin webmin and usermin from all smartphones, tablets, and desktops.

**Comment by chris001** [comment]
Here's some resources to make it easy for when it's time to develop a JSON REST web service API for the webmin core app, and a JSON REST client for the theme app.
* https://www.getpostman.com/apps
* Demo of how easy a REST API is to develop live using this Postman app https://www.youtube.com/watch?v=11tI5owAZhg

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
